modules/optimization/src/executioners/TransientAndAdjoint.C

```diff
@@ -0,0 +1,113 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "TransientAndAdjoint.h"
+#include "OptimizationAppTypes.h"
+
+registerMooseObject("OptimizationApp", TransientAndAdjoint);
+
+InputParameters
+TransientAndAdjoint::validParams()
+{
+  InputParameters params = Transient::validParams();
+  params += AdjointTransientSolve::validParams();
+  params.addClassDescription("Executioner for evaluating transient simulations and their adjoint.");
+
+  // We need to full matrix for the adjoint solve, so set this to NEWTON
+  params.set<MooseEnum>("solve_type") = "newton";
+  params.suppressParameter<MooseEnum>("solve_type");
+
+  return params;
+}
+
+TransientAndAdjoint::TransientAndAdjoint(const InputParameters & parameters)
+  : Transient(parameters),
+    _adjoint_solve(*this),
+    _forward_times(declareRecoverableData<std::vector<Real>>("forward_times"))
+{
+  // Error out on unsupported time integration schemes
+  switch (getTimeScheme())
+  {
+    case Moose::TI_IMPLICIT_EULER:
+      break;
+    default:
+      paramError(
+          "scheme", getParam<MooseEnum>("scheme"), " is not supported for computing adjoint.");
+      break;
+  }
+}
+
+void
+TransientAndAdjoint::preExecute()
+{
+  Transient::preExecute();
+
+  // Save the forward initial condition
+  _adjoint_solve.insertForwardSolution(_t_step);
+
+  // Save the time of the initial condition
+  // The vector of times should be reset unless we are recovering
+  if (!_app.isRecovering())
+    _forward_times = {_time};
+  else
+    _forward_times.push_back(_time);
+}
+
+void
+TransientAndAdjoint::postStep()
+{
+  Transient::postStep();
+
+  // Save the converged forward solution and time
+  if (lastSolveConverged())
+  {
+    _adjoint_solve.insertForwardSolution(_t_step);
+    _forward_times.push_back(_time);
+  }
+}
+
+void
+TransientAndAdjoint::postExecute()
+{
+  Transient::postExecute();
+
+  // If it is a half transient, then the app is meant to be run with recovery. Therefore, it doesn't
+  // make sense to run the adjoint calculation since we aren't getting to the final time required
+  // for a consistent adjoint.
+  if (_app.halfTransient())
```

Comment:
So I do want to solve the adjoint equation when recovering. I just don't want to solve it when it is checkpointed during the first half transient. In reality, it does work properly without this `return`, but there is a CSVDiff since the output from the adjoint solve in Part1 is not overridden when recovering. I think it's the same reason why we don't output on `FINAL` if it's a half transient.

Reply:
So this is a workaround for an underlying issue in CSVDiff? Like we need to be appending to the CSVDiff?

Reply:
I'm really unsure if it's even an issue. The resulting CSV makes sense, there are rows from the adjoint solve representing the half transient, and there are rows representing the full transient. The problem is the gold file doesn't contain the half transient entries. I'm don't know how we would change CSVDiff to support this, maybe having an option to only compare certain rows? But conceptually, I don't think it makes sense to even solve the adjoint problem during a half transient, which is why I have the `return`.

Reply:
Alrighty